In [82]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from geopy.geocoders import Nominatim
import requests
import googlemaps
import json
import sys
import re

<h2>Things to do:</h2>
<li>Research all columns and their properties</li>
<li>Fix the null values. We can use either imputation or research to see what can be replaced with.</li>
<li>Check the distribution and outlier values</li>
<li>Rough idea of data analysis: There is no input-output mapping/learning for this data as such since all of them are related to accidents. The most possible option would be clustering based on low-risk, medium-risk, high-risk accident areas. A person walking or driving can enter the location of the current place or ask app to track their location. Based on location such as area name, or coordinates, cluster the location into the three categories. A possible option would be to have a heatmap overlayed on top of normal map data where different color (green, yellow, red) can be used to denote different </li>

<h3>Null/Missing values</h3>
<li>locations fields with missing data indicated as (0, 0)</li>
<li>address fields </li>

In [3]:
df = pd.read_csv('../data/Traffic_Collision_Data_from_2010_to_Present.csv')
df

DR Number Date Reported Date Occurred  Time Occurred  Area ID  \
0       190319651    08/24/2019    08/24/2019            450        3   
1       190319680    08/30/2019    08/30/2019           2320        3   
2       190413769    08/25/2019    08/25/2019            545        4   
3       190127578    11/20/2019    11/20/2019            350        1   
4       190319695    08/30/2019    08/30/2019           2100        3   
...           ...           ...           ...            ...      ...   
570367  222005125    01/29/2022    01/28/2022           2245       20   
570368  221504553    01/15/2022    01/14/2022           1553       15   
570369  221504766    01/20/2022    01/20/2022           1720       15   
570370  221505127    01/28/2022    01/28/2022           1430       15   
570371  220805005    01/26/2022    01/25/2022           2205        8   

          Area Name  Reporting District  Crime Code Crime Code Description  \
0         Southwest                 356         997      TRAFFIC COLLISION   
1         Southwest                 355         997      TRAFFIC COLLISION   
2        Hollenbeck                 422         997      TRAFFIC COLLISION   
3           Central                 128         997      TRAFFIC COLLISION   
4         Southwest                 374         997      TRAFFIC COLLISION   
...             ...                 ...         ...                    ...   
570367      Olympic                2013         997      TRAFFIC COLLISION   
570368  N Hollywood                1566         997      TRAFFIC COLLISION   
570369  N Hollywood                1547         997      TRAFFIC COLLISION   
570370  N Hollywood                1547         997      TRAFFIC COLLISION   
570371      West LA                 885         997      TRAFFIC COLLISION   

                                  MO Codes  Victim Age Victim Sex  \
0                 3036 3004 3026 3101 4003        22.0          M   
1       3037 3006 3028 3030 3039 3101 4003        30.0          F   
2                 3101 3401 3701 3006 3030         NaN          M   
3            0605 3101 3401 3701 3011 3034        21.0          M   
4            0605 4025 3037 3004 3025 3101        49.0          M   
...                                    ...         ...        ...   
570367            1402 3003 3024 3029 4027        50.0          M   
570368       3004 3026 4015 3036 3038 3101        28.0          M   
570369  3002 3101 4026 3011 3026 3032 3037        21.0          M   
570370       3006 3028 4026 3033 3034 3037        99.0          X   
570371                                4027        34.0          F   

       Victim Descent  Premise Code Premise Description  \
0                   H         101.0              STREET   
1                   H         101.0              STREET   
2                   X         101.0              STREET   
3                   H         101.0              STREET   
4                   B         101.0              STREET   
...               ...           ...                 ...   
570367              O         101.0              STREET   
570368              W         101.0              STREET   
570369              O         101.0              STREET   
570370              X         101.0              STREET   
570371              O         101.0              STREET   

                                Address                        Cross Street  \
0       JEFFERSON                    BL     NORMANDIE                    AV   
1       JEFFERSON                    BL                          W  WESTERN   
2                           N  BROADWAY  W  EASTLAKE                     AV   
3                                   1ST                             CENTRAL   
4                 MARTIN LUTHER KING JR     ARLINGTON                    AV   
...                                 ...                                 ...   
570367  3RD                          ST     HOBART                       BL   
570368  VINELAND        

In [6]:
df['Crime Code Description'].value_counts()

TRAFFIC COLLISION    570372
Name: Crime Code Description, dtype: int64

<h2>Potential features that can be used for clustering model</h2>
<li>Time Occurred: We can create some features such part of day the accident occurred (morning, evening, night)</li>
<li>Area ID</li>
<li>Area Name, Address, and Cross Street: Combined Together</li>
<li>Reporting District</li>
<li>MO Codes (research it)</li>
<li>Premise Code (research it)</li>
<li>Premise Description</li>
<li>Location</li>

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 570372 entries, 0 to 570371
Data columns (total 18 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   DR Number               570372 non-null  int64  
 1   Date Reported           570372 non-null  object 
 2   Date Occurred           570372 non-null  object 
 3   Time Occurred           570372 non-null  int64  
 4   Area ID                 570372 non-null  int64  
 5   Area Name               570372 non-null  object 
 6   Reporting District      570372 non-null  int64  
 7   Crime Code              570372 non-null  int64  
 8   Crime Code Description  570372 non-null  object 
 9   MO Codes                483387 non-null  object 
 10  Victim Age              487109 non-null  float64
 11  Victim Sex              561234 non-null  object 
 12  Victim Descent          560341 non-null  object 
 13  Premise Code            569410 non-null  float64
 14  Premise Description 

<h3>List of columns that have null values</h3>


In [18]:
df.isnull().sum()

DR Number                     0
Date Reported                 0
Date Occurred                 0
Time Occurred                 0
Area ID                       0
Area Name                     0
Reporting District            0
Crime Code                    0
Crime Code Description        0
MO Codes                  86985
Victim Age                83263
Victim Sex                 9138
Victim Descent            10031
Premise Code                962
Premise Description         963
Address                       0
Cross Street              27224
Location                      0
dtype: int64

<h4>For this notebook, I will be doing the following:</h4>
<li>Fill null values for 'location coordinates', 'cross street'</li>
<li>Combine certain columns together for conciseness</li>
<li>Feature creation for certain columns.</li>
<li>Uploading JSON file on FireBase for access and this other stuff on GitHub.</li>

<p><b># Case 1: Area Name, Address and Cross Street -> Use all to find coordinates</b></p>

In [30]:
df[['Area Name', 'Address', 'Cross Street', 'Location']][(df['Cross Street'].notna()) & (df['Location'] == '(0.0, 0.0)') & (df['Address'] != '00')]

Area Name                          Address  \
2598       Van Nuys                        RIVERSIDE   
2622    West Valley  VENTURA                      BL   
2651    West Valley  VENTURA                      BL   
2769       Van Nuys  VAN NUYS                     BL   
2960    West Valley  RESEDA                       BL   
...             ...                              ...   
567954    Southeast  COMPTON                      AV   
567959     Foothill  SUTTER                       AV   
567962      Topanga  VENTURA                      BL   
567963  N Hollywood  LANKERSHIM                   BL   
567965      Mission  FRANK MODUGN                 DR   

                           Cross Street    Location  
2598                             FULTON  (0.0, 0.0)  
2622    LA MAIDA                     ST  (0.0, 0.0)  
2651    DONNA                        AV  (0.0, 0.0)  
2769    HUSTON                       ST  (0.0, 0.0)  
2960    VICTORY                      BL  (0.0, 0.0)  
...                                 ...         ...  
567954  108TH                        ST  (0.0, 0.0)  
567959  VAN NUYS                     BL  (0.0, 0.0)  
567962  ALHAMA                       DR  (0.0, 0.0)  
567963  KITTRIDGE                    ST  (0.0, 0.0)  
567965  HUBBARD                      AV  (0.0, 0.0)  

[733 rows x 4 columns]

In [4]:
df1 = df[(df['Cross Street'].notna()) & (df['Location'] == '(0.0, 0.0)') & (df['Address'] != '00')]

<p><b># Case 2: Area Name and Address given, but no cross street -> Use only Area Name and Address to find coordinates</b></p>

In [31]:
df[['Area Name', 'Address', 'Cross Street', 'Location']][(df['Cross Street'].isnull()) & (df['Location'] == '(0.0, 0.0)') & (df['Address'] != '00')]

Area Name                                   Address Cross Street  \
17957       Mission   8100    ORION                        AV          NaN   
20034      Van Nuys  13300    VALLEYHEART                  DR          NaN   
23816       Topanga  21200    KITTRIDGE                    ST          NaN   
24785    Devonshire  20300    CORALINE                     CI          NaN   
25212    Devonshire  18500    SUNBURST                     ST          NaN   
26206     Southeast  10300    HOLMES                       AV          NaN   
31487   N Hollywood   6500    RIVERTON                     AV          NaN   
42393   N Hollywood  12200    COHASSET                     ST          NaN   
50452    Devonshire   9300    TAMPA                        AV          NaN   
51675      Van Nuys   7300    HAZELTINE                    AV          NaN   
52455      Van Nuys   7100    HAZELTINE                    AV          NaN   
52914   West Valley   7200    RESEDA                       BL          NaN   
54126        Harbor   1200 W  7TH                          ST          NaN   
55351       Mission   7800    NOLLAN                       PL          NaN   
56959       Mission  13700    SIMSHAW                      AV          NaN   
58090   77th Street   3400 W  78TH                         ST          NaN   
58697   77th Street    800 W  51ST                         ST          NaN   
64803       Topanga   7200    ALABAMA                      AV          NaN   
65536     Southeast   9500    COMPTON                      AV          NaN   
173945    Southwest   3800 S  HOBART                       BL          NaN   
179940    Hollywood   5600    FERNWOOD                     AV          NaN   
180312    Hollywood   1400 N  CHEROKEE                     AV          NaN   
203717     Foothill  12700    VAN NUYS                     BL          NaN   
208675      Mission  14100    POLK                         ST          NaN   
293077  N Hollywood                        3400    CAHUENGA W          NaN   
306599    Northeast   2200    YOSEMITY                     DR          NaN   
547000  77th Street   7900 S  VERMONT                      AV          NaN   
547077  77th Street    800 W  54TH                         ST          NaN   
555917   Devonshire  10800    BALBOA                       BL          NaN   
562900  N Hollywood   4300    COLDWATER CANYON             AV          NaN   
564701    Southwest   3500 S  WESTERN                      AV          NaN   
565178       Newton   3800 S  SAN PEDRO                    ST          NaN   
567964      Olympic   3400 W  3RD                          ST          NaN   

          Location  
17957   (0.0, 0.0)  
20034   (0.0, 0.0)  
23816   (0.0, 0.0)  
24785   (0.0, 0.0)  
25212   (0.0, 0.0)  
26206   (0.0, 0.0)  
31487   (0.0, 0.0)  
42393   (0.0, 0.0)  
50452   (0.0, 0.0)  
51675   (0.0, 0.0)  
52455   (0.0, 0.0)  
52914   (0.0, 0.0)  
54126   (0.0, 0.0)  
55351   (0.0, 0.0)  
56959   (0.0, 0.0)  
58090   (0.0, 0.0)  
58697   (0.0, 0.0)  
64803   (0.0, 0.0)  
65536   (0.0, 0.0)  
173945  (0.0, 0.0)  
179940  (0.0, 0.0)  
180312  (0.0, 0.0)  
203717  (0.0, 0.0)  
208675  (0.0, 0.0)  
293077  (0.0, 0.0)  
306599  (0.0, 0.0)  
547000  (0.0, 0.0)  
547077  (0.0, 0.0)  
555917  (0.0, 0.0)  
562900  (0.0, 0.0)  
564701  (0.0, 0.0)  
565178  (0.0, 0.0)  
567964  (0.0, 0.0)

In [5]:
df2 = df[(df['Cross Street'].isnull()) & (df['Location'] == '(0.0, 0.0)') & (df['Address'] != '00')]

<p><b># Case 3: Area Name given, but no address and cross street  -> Use only Area Name to find coordinates</b></p>

In [32]:
df[['Area Name', 'Address', 'Cross Street', 'Location']][(df['Cross Street'].isnull()) & (df['Location'] == '(0.0, 0.0)') & (df['Address'] == '00')]

Area Name Address Cross Street    Location
2198     Devonshire      00          NaN  (0.0, 0.0)
2201     Devonshire      00          NaN  (0.0, 0.0)
2207     Devonshire      00          NaN  (0.0, 0.0)
2285        Topanga      00          NaN  (0.0, 0.0)
2294        Topanga      00          NaN  (0.0, 0.0)
2295        Topanga      00          NaN  (0.0, 0.0)
2297        Topanga      00          NaN  (0.0, 0.0)
2298        Topanga      00          NaN  (0.0, 0.0)
2299        Topanga      00          NaN  (0.0, 0.0)
2834         Harbor      00          NaN  (0.0, 0.0)
4312     Devonshire      00          NaN  (0.0, 0.0)
7198    West Valley      00          NaN  (0.0, 0.0)
47922    Devonshire      00          NaN  (0.0, 0.0)
48016    Devonshire      00          NaN  (0.0, 0.0)
48173    Devonshire      00          NaN  (0.0, 0.0)
48314    Devonshire      00          NaN  (0.0, 0.0)
48323       Topanga      00          NaN  (0.0, 0.0)
303112      Olympic      00          NaN  (0.0, 0.0)
303882      Olympic      00          NaN  (0.0, 0.0)

In [6]:
df3 = df[(df['Cross Street'].isnull()) & (df['Location'] == '(0.0, 0.0)') & (df['Address'] == '00')]

<p><b># Case 4: Area Name and Address and coordinates given, but no cross street -> Use coordinates to find the cross street</b></p>

In [49]:
df[['Area Name', 'Address', 'Cross Street', 'Location']][(df['Cross Street'].isnull()) & (df['Location'] != '(0.0, 0.0)')]

Area Name                                   Address Cross Street  \
9        Hollenbeck   3500    VALLEY                       BL          NaN   
19           Harbor  25800 S  VERMONT                      AV          NaN   
26           Harbor    200 W  14TH                         ST          NaN   
107         West LA   1000    BROXTON                      AV          NaN   
174       Northeast   3800 W  SUNSET                       BL          NaN   
...             ...                                       ...          ...   
570233      West LA    600    HAMPDEN                      PL          NaN   
570236      Pacific   6900    LA TIJERA                    BL          NaN   
570276       Newton   1900 S  LOS ANGELES                  ST          NaN   
570277  77th Street                          7600 S  BROADWAY          NaN   
570365     Van Nuys   5700 N  SEPULVEDA                    BL          NaN   

                    Location  
9         (34.07, -118.1998)  
19      (33.7894, -118.2936)  
26      (33.7315, -118.2814)  
107      (34.0625, -118.447)  
174     (34.0917, -118.2741)  
...                      ...  
570233   (34.042, -118.5273)  
570236  (33.9757, -118.3737)  
570276  (34.0304, -118.2637)  
570277  (33.9703, -118.2783)  
570365   (34.174, -118.4661)  

[27172 rows x 4 columns]

In [42]:
df[['Area Name', 'Address', 'Cross Street', 'Location']][df['Location'] == '(0.0, 0.0)']

Area Name                                  Address  \
2198     Devonshire                                       00   
2201     Devonshire                                       00   
2207     Devonshire                                       00   
2285        Topanga                                       00   
2294        Topanga                                       00   
...             ...                                      ...   
567959     Foothill          SUTTER                       AV   
567962      Topanga          VENTURA                      BL   
567963  N Hollywood          LANKERSHIM                   BL   
567964      Olympic  3400 W  3RD                          ST   
567965      Mission          FRANK MODUGN                 DR   

                           Cross Street    Location  
2198                                NaN  (0.0, 0.0)  
2201                                NaN  (0.0, 0.0)  
2207                                NaN  (0.0, 0.0)  
2285                                NaN  (0.0, 0.0)  
2294                                NaN  (0.0, 0.0)  
...                                 ...         ...  
567959  VAN NUYS                     BL  (0.0, 0.0)  
567962  ALHAMA                       DR  (0.0, 0.0)  
567963  KITTRIDGE                    ST  (0.0, 0.0)  
567964                              NaN  (0.0, 0.0)  
567965  HUBBARD                      AV  (0.0, 0.0)  

[785 rows x 4 columns]

In [7]:
df4 = df[df['Location'] != '(0.0, 0.0)']

In [8]:
print(f'Sum of all elements having missing coordinates: {len(df1) + len(df2) + len(df3)}')

Sum of all elements having missing coordinates: 785


<p>Since there's no coordinates for these area names, we can use a geolocation API to convert addresses into coordinates. We can use the below API to get the coordinates</p>
<li>Nominatim, OpenStreetMap API — (free)</li>
<li><a>https://towardsdatascience.com/pythons-geocoding-convert-a-list-of-addresses-into-a-map-f522ef513fd6</a></li>
<li><a>https://programminghistorian.org/en/lessons/mapping-with-python-leaflet</a></li>

geolocator = Nominatim(user_agent="example app")

cross street, area name, address 
 
geolocator.geocode("West Valley, La Maida St, Los Angeles, California").point

lat = geolocator.geocode("Topanga, Ventura Boulevard, Los Angeles, California").point[0]
lon = geolocator.geocode("Topanga, Ventura Boulevard, Los Angeles, California").point[1]
print(f'latitude:{lat}, longitude:{lon}')
geolocator.reverse('34.07, -118.1998')


<p>-----------------------------------------------------------------------------------------------------------------------------------------</p>
<b>Order of addressing</b>
<p>Area Name: Neighborhood</p>
<p>Address: Street Name</p>
<p>Cross Street: Intersecting Street</p>
<p>City: Los Angeles</p>
<p>State: California</p>

df1

In [9]:
location_ = []
for index, row in df1.iterrows():
    address = str(" ".join(row['Address'].split()))
    cross_street = str(" ".join(row['Cross Street'].split()))
    area = str(" ".join(row['Area Name'].split()))
    str_location = address + ',' + area + ',' + cross_street + ',' + 'Los Angeles' + ',' + 'California'
    location_.append(str_location)


df1['location_combined'] = location_

/var/folders/gc/h6mhk26x34vc_b1d1db5rxh40000gn/T/ipykernel_94530/94293974.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['location_combined'] = location_


df2

In [20]:
location_ = []
for index, row in df2.iterrows():
    address = str(" ".join(row['Address'].split()))
    area = str(" ".join(row['Area Name'].split()))
    str_location = address + ',' + area + ',' + 'Los Angeles' + ',' + 'California'
    location_.append(str_location)

df2['location_combined'] = location_

/var/folders/gc/h6mhk26x34vc_b1d1db5rxh40000gn/T/ipykernel_94530/2202953927.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['location_combined'] = location_


df3

In [24]:
location_ = []
for index, row in df3.iterrows():
    area = str(" ".join(row['Area Name'].split()))
    str_location = area + ',' + 'Los Angeles' + ',' + 'California'
    location_.append(str_location)

df3['location_combined'] = location_

/var/folders/gc/h6mhk26x34vc_b1d1db5rxh40000gn/T/ipykernel_94530/2247413596.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df3['location_combined'] = location_


<p>-----------------------------------------------------------------------------------------------------------------------------------------</p>
<b>Order of addressing</b>
<p>Area Name: Neighborhood</p>
<p>Address: Street Name</p>
<p>Cross Street: Intersecting Street</p>
<p>City: Los Angeles</p>
<p>State: California</p>

In [12]:
# Initializing the API using credentials
gmaps_key = googlemaps.Client(key="AIzaSyBJ66MFlaEB_SV5tZJk3UjUBv006XEceQs")

In [13]:
def geocode(address):
    g = gmaps_key.geocode(address)
    if len(g) != 0:
        lat = g[0]["geometry"]["location"]["lat"]
        long = g[0]["geometry"]["location"]["lng"]
    
    else:
        return('-1')

    return (lat, long) 

In [27]:
df1['Location'] = df1['location_combined'].apply(lambda x: geocode(x))
print('Done with df1')
df2['Location'] = df2['location_combined'].apply(lambda x: geocode(x))
print('Done with df2')
df3['Location'] = df3['location_combined'].apply(lambda x: geocode(x))
print('Done with df3')

/var/folders/gc/h6mhk26x34vc_b1d1db5rxh40000gn/T/ipykernel_94530/561757893.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['Location'] = df1['location_combined'].apply(lambda x: geocode(x))


Done with df1


/var/folders/gc/h6mhk26x34vc_b1d1db5rxh40000gn/T/ipykernel_94530/561757893.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['Location'] = df2['location_combined'].apply(lambda x: geocode(x))


Done with df2
Done with df3


/var/folders/gc/h6mhk26x34vc_b1d1db5rxh40000gn/T/ipykernel_94530/561757893.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df3['Location'] = df3['location_combined'].apply(lambda x: geocode(x))


In [28]:
df1[['location_combined', 'Location']]

location_combined  \
2598     RIVERSIDE,Van Nuys,FULTON,Los Angeles,California   
2622    VENTURA BL,West Valley,LA MAIDA ST,Los Angeles...   
2651    VENTURA BL,West Valley,DONNA AV,Los Angeles,Ca...   
2769    VAN NUYS BL,Van Nuys,HUSTON ST,Los Angeles,Cal...   
2960    RESEDA BL,West Valley,VICTORY BL,Los Angeles,C...   
...                                                   ...   
567954  COMPTON AV,Southeast,108TH ST,Los Angeles,Cali...   
567959  SUTTER AV,Foothill,VAN NUYS BL,Los Angeles,Cal...   
567962  VENTURA BL,Topanga,ALHAMA DR,Los Angeles,Calif...   
567963  LANKERSHIM BL,N Hollywood,KITTRIDGE ST,Los Ang...   
567965  FRANK MODUGN DR,Mission,HUBBARD AV,Los Angeles...   

                          Location  
2598      (34.157936, -118.448372)  
2622    (34.1586274, -118.4312574)  
2651    (34.2079816, -118.5483412)  
2769     (34.1594324, -118.420345)  
2960    (34.2120269, -118.5360834)  
...                            ...  
567954  (33.9743629, -118.2476761)  
567959  (34.2158867, -118.4487664)  
567962  (34.1581991, -118.5932248)  
567963    (34.190315, -118.396532)  
567965  (34.2931621, -118.4497882)  

[733 rows x 2 columns]

In [29]:
df2[['location_combined', 'Location']]

location_combined  \
17957        8100 ORION AV,Mission,Los Angeles,California   
20034   13300 VALLEYHEART DR,Van Nuys,Los Angeles,Cali...   
23816   21200 KITTRIDGE ST,Topanga,Los Angeles,California   
24785   20300 CORALINE CI,Devonshire,Los Angeles,Calif...   
25212   18500 SUNBURST ST,Devonshire,Los Angeles,Calif...   
26206    10300 HOLMES AV,Southeast,Los Angeles,California   
31487   6500 RIVERTON AV,N Hollywood,Los Angeles,Calif...   
42393   12200 COHASSET ST,N Hollywood,Los Angeles,Cali...   
50452     9300 TAMPA AV,Devonshire,Los Angeles,California   
51675   7300 HAZELTINE AV,Van Nuys,Los Angeles,California   
52455   7100 HAZELTINE AV,Van Nuys,Los Angeles,California   
52914   7200 RESEDA BL,West Valley,Los Angeles,California   
54126         1200 W 7TH ST,Harbor,Los Angeles,California   
55351       7800 NOLLAN PL,Mission,Los Angeles,California   
56959     13700 SIMSHAW AV,Mission,Los Angeles,California   
58090   3400 W 78TH ST,77th Street,Los Angeles,California   
58697    800 W 51ST ST,77th Street,Los Angeles,California   
64803      7200 ALABAMA AV,Topanga,Los Angeles,California   
65536    9500 COMPTON AV,Southeast,Los Angeles,California   
173945  3800 S HOBART BL,Southwest,Los Angeles,California   
179940  5600 FERNWOOD AV,Hollywood,Los Angeles,California   
180312  1400 N CHEROKEE AV,Hollywood,Los Angeles,Calif...   
203717  12700 VAN NUYS BL,Foothill,Los Angeles,California   
208675       14100 POLK ST,Mission,Los Angeles,California   
293077  3400 CAHUENGA W,N Hollywood,Los Angeles,Califo...   
306599  2200 YOSEMITY DR,Northeast,Los Angeles,California   
547000  7900 S VERMONT AV,77th Street,Los Angeles,Cali...   
547077   800 W 54TH ST,77th Street,Los Angeles,California   
555917  10800 BALBOA BL,Devonshire,Los Angeles,California   
562900  4300 COLDWATER CANYON AV,N Hollywood,Los Angel...   
564701  3500 S WESTERN AV,Southwest,Los Angeles,Califo...   
565178  3800 S SAN PEDRO ST,Newton,Los Angeles,California   
567964       3400 W 3RD ST,Olympic,Los Angeles,California   

                                 Location  
17957           (34.2172329, -118.472074)  
20034          (34.1515185, -118.4228472)  
23816          (34.1903549, -118.5937491)  
24785          (34.2572761, -118.5285392)  
25212          (34.2345893, -118.5372391)  
26206          (33.9428391, -118.2412382)  
31487           (34.188655, -118.3657146)  
42393   (34.20656820000001, -118.3988738)  
50452           (34.257574, -118.5532336)  
51675           (34.2030596, -118.439744)  
52455          (34.1993616, -118.4399937)  
52914          (34.2013168, -118.5359368)  
54126          (33.7379088, -118.3033795)  
55351          (34.2146686, -118.4451056)  
56959           (34.321286, -118.4151008)  
58090          (33.9690848, -118.3308676)  
58697           (33.9964453, -118.287144)  
64803          (34.2013656, -118.5990452)  
65536          (33.9506787, -118.2461166)  
173945         (34.0174105, -118.3075188)  
179940         (34.0955019, -118.3111033)  
180312         (34.0961722, -118.3352797)  
203717  (34.27470479999999, -118.4097154)  
208675         (34.3155639, -118.4408072)  
293077         (34.1319869, -118.3518106)  
306599         (34.1351578, -118.2153329)  
547000         (33.9676443, -118.2915865)  
547077          (33.9929746, -118.287332)  
555917         (34.2661949, -118.5019753)  
562900         (34.1482256, -118.4137951)  
564701         (34.0243604, -118.3086194)  
565178         (34.0116851, -118.2673521)  
567964         (34.0685541, -118.2888849)

In [30]:
df3[['location_combined', 'Location']]

location_combined                    Location
2198     Devonshire,Los Angeles,California  (34.2572761, -118.5285392)
2201     Devonshire,Los Angeles,California  (34.2572761, -118.5285392)
2207     Devonshire,Los Angeles,California  (34.2572761, -118.5285392)
2285        Topanga,Los Angeles,California  (34.0919452, -118.6021321)
2294        Topanga,Los Angeles,California  (34.0919452, -118.6021321)
2295        Topanga,Los Angeles,California  (34.0919452, -118.6021321)
2297        Topanga,Los Angeles,California  (34.0919452, -118.6021321)
2298        Topanga,Los Angeles,California  (34.0919452, -118.6021321)
2299        Topanga,Los Angeles,California  (34.0919452, -118.6021321)
2834         Harbor,Los Angeles,California    (33.779981, -118.287743)
4312     Devonshire,Los Angeles,California  (34.2572761, -118.5285392)
7198    West Valley,Los Angeles,California  (34.1849537, -118.6032508)
47922    Devonshire,Los Angeles,California  (34.2572761, -118.5285392)
48016    Devonshire,Los Angeles,California  (34.2572761, -118.5285392)
48173    Devonshire,Los Angeles,California  (34.2572761, -118.5285392)
48314    Devonshire,Los Angeles,California  (34.2572761, -118.5285392)
48323       Topanga,Los Angeles,California  (34.0919452, -118.6021321)
303112      Olympic,Los Angeles,California  (34.0522342, -118.2436849)
303882      Olympic,Los Angeles,California  (34.0522342, -118.2436849)

In [32]:
df1.drop(columns=['location_combined'], inplace=True)
df2.drop(columns=['location_combined'], inplace=True)
df3.drop(columns=['location_combined'], inplace=True)

/opt/anaconda3/lib/python3.9/site-packages/pandas/core/frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [35]:
frames = [df1,df2,df3,df4]
df_final = pd.concat(frames, ignore_index=True)

In [40]:
df_final.columns

Index(['DR Number', 'Date Reported', 'Date Occurred', 'Time Occurred',
       'Area ID', 'Area Name', 'Reporting District', 'Crime Code',
       'Crime Code Description', 'MO Codes', 'Victim Age', 'Victim Sex',
       'Victim Descent', 'Premise Code', 'Premise Description', 'Address',
       'Cross Street', 'Location'],
      dtype='object')

In [78]:
df_final.drop(columns=["Crime Code", "Crime Code Description", "Victim Age", "Victim Sex", "Victim Descent", "MO Codes", "Premise Code", "Reporting District"], inplace=True)

In [70]:
area_ids = sorted(list(df_final['Area ID'].value_counts().index))

In [71]:
# Storing areas by their IDs
area_by_ID = {}
for area_id in area_ids:
    area_by_ID[area_id] = df_final['Area Name'][df_final['Area ID'] == area_id].iloc[0]

In [72]:
area_by_ID

{1: 'Central',
 2: 'Rampart',
 3: 'Southwest',
 4: 'Hollenbeck',
 5: 'Harbor',
 6: 'Hollywood',
 7: 'Wilshire',
 8: 'West LA',
 9: 'Van Nuys',
 10: 'West Valley',
 11: 'Northeast',
 12: '77th Street',
 13: 'Newton',
 14: 'Pacific',
 15: 'N Hollywood',
 16: 'Foothill',
 17: 'Devonshire',
 18: 'Southeast',
 19: 'Mission',
 20: 'Olympic',
 21: 'Topanga'}

In [36]:
import pandas as pd
import re
df_final = pd.read_csv('../data/df_modified_ver1.csv')
df_final.set_index('Unnamed: 0', inplace=True)
df_final

DR Number Date Reported Date Occurred  Time Occurred  Area ID  \
Unnamed: 0                                                                  
0           200908529    04/16/2020    04/16/2020           1740        9   
1           201008046    04/14/2020    04/14/2020           1450       10   
2           201008099    04/16/2020    04/16/2020           1250       10   
3           200908540    04/16/2020    04/16/2020           1515        9   
4           201007987    04/11/2020    04/11/2020           2330       10   
...               ...           ...           ...            ...      ...   
570367      222005125    01/29/2022    01/28/2022           2245       20   
570368      221504553    01/15/2022    01/14/2022           1553       15   
570369      221504766    01/20/2022    01/20/2022           1720       15   
570370      221505127    01/28/2022    01/28/2022           1430       15   
570371      220805005    01/26/2022    01/25/2022           2205        8   

              Area Name Premise Description                          Address  \
Unnamed: 0                                                                     
0              Van Nuys              STREET                        RIVERSIDE   
1           West Valley              STREET  VENTURA                      BL   
2           West Valley              STREET  VENTURA                      BL   
3              Van Nuys              STREET  VAN NUYS                     BL   
4           West Valley              STREET  RESEDA                       BL   
...                 ...                 ...                              ...   
570367          Olympic              STREET  3RD                          ST   
570368      N Hollywood              STREET  VINELAND                     AV   
570369      N Hollywood              STREET  BURBANK                      BL   
570370      N Hollywood              STREET  MAGNOLIA                     BL   
570371          West LA              STREET                         NATIONAL   

                               Cross Street                    Location  
Unnamed: 0                                                               
0                                    FULTON    (34.157936, -118.448372)  
1           LA MAIDA                     ST  (34.1586274, -118.4312574)  
2           DONNA                        AV  (34.2079816, -118.5483412)  
3           HUSTON                       ST   (34.1594324, -118.420345)  
4           VICTORY                      BL  (34.2120269, -118.5360834)  
...                                     ...                         ...  
570367      HOBART                       BL         (34.069, -118.3054)  
570368      RIVERSIDE                    DR        (34.1524, -118.3703)  
570369      TROOST                       AV        (34.1722, -118.3866)  
570370      HOLLYWOOD                    FY        (34.1649, -118.3825)  
570371                               I10FWY        (34.0405, -118.4139)  

[570372 rows x 10 columns]

In [37]:
import re
def time_extractor(h):
    x = str(h)

    if len(x) == 4:
        return int(re.search(r'([0-9]{2})', x).group())
    else:
        return int(re.search(r'([0-9]{1})', x).group())


In [38]:
df_final['month'] = df_final['Date Occurred'].apply(lambda x: int(str(re.search('[0-9]{2}', x).group())))
df_final['day'] = df_final['Date Occurred'].apply(lambda x: int(str(re.search('/[0-9]{2}/', x).group()).replace('/', '')))
df_final['year'] = df_final['Date Occurred'].apply(lambda x: int(str(re.search('/[0-9]{4}', x).group()).replace('/', '')))
df_final['hour'] = df_final['Time Occurred'].apply(lambda x: time_extractor(x))

In [39]:
def day_category(num):
    if 6 <= num < 12:
        return 'morning'
    elif 12 <= num < 16:
        return 'afternoon'
    elif 16 <= num < 19:
        return 'evening'
    elif 19 <= num <= 23 or 1 <= num < 6:
        return 'night'


df_final['day_category'] =  df_final['hour'].apply(lambda x: day_category(x))


In [42]:
df_final.head()

DR Number Date Reported Date Occurred  Time Occurred  Area ID  \
Unnamed: 0                                                                  
0           200908529    04/16/2020    04/16/2020           1740        9   
1           201008046    04/14/2020    04/14/2020           1450       10   
2           201008099    04/16/2020    04/16/2020           1250       10   
3           200908540    04/16/2020    04/16/2020           1515        9   
4           201007987    04/11/2020    04/11/2020           2330       10   

              Area Name Premise Description                          Address  \
Unnamed: 0                                                                     
0              Van Nuys              STREET                        RIVERSIDE   
1           West Valley              STREET  VENTURA                      BL   
2           West Valley              STREET  VENTURA                      BL   
3              Van Nuys              STREET  VAN NUYS                     BL   
4           West Valley              STREET  RESEDA                       BL   

                               Cross Street                    Location  \
Unnamed: 0                                                                
0                                    FULTON    (34.157936, -118.448372)   
1           LA MAIDA                     ST  (34.1586274, -118.4312574)   
2           DONNA                        AV  (34.2079816, -118.5483412)   
3           HUSTON                       ST   (34.1594324, -118.420345)   
4           VICTORY                      BL  (34.2120269, -118.5360834)   

            month  day  year  hour day_category  
Unnamed: 0                                       
0               4   16  2020    17      evening  
1               4   14  2020    14    afternoon  
2               4   16  2020    12    afternoon  
3               4   16  2020    15    afternoon  
4               4   11  2020    23        night

In [45]:
df_final.drop(columns=['Date Reported', 'Date Occurred', 'Time Occurred', 'Address', 'Cross Street', 'Area Name'], inplace=True)

In [46]:
df_final.head()

DR Number  Area ID Premise Description  \
Unnamed: 0                                           
0           200908529        9              STREET   
1           201008046       10              STREET   
2           201008099       10              STREET   
3           200908540        9              STREET   
4           201007987       10              STREET   

                              Location  month  day  year  hour day_category  
Unnamed: 0                                                                   
0             (34.157936, -118.448372)      4   16  2020    17      evening  
1           (34.1586274, -118.4312574)      4   14  2020    14    afternoon  
2           (34.2079816, -118.5483412)      4   16  2020    12    afternoon  
3            (34.1594324, -118.420345)      4   16  2020    15    afternoon  
4           (34.2120269, -118.5360834)      4   11  2020    23        night

In [47]:
df_final.to_csv(r'../data/df_modified_ver2.csv')

In [48]:
import pandas as pd
import requests
import json
import sys

# Function to convert CSV file to JSON
def make_json(csvFilePath, jsonFilePath):

    #1. Import data into pandas
    accident_data = pd.read_csv(csvFilePath)
    accident_data.set_index('Unnamed: 0', inplace=True)
    accident_data = accident_data.fillna('')
    #3. Sort by customer id
    accident_data.sort_values(by=['DR Number'], inplace=True)
    cols = list(accident_data.columns)
    cols.remove('DR Number')

    #4. Convert CSV file to JSON by making dictionary first and then writing to json file

    data = {}

    for index, row in accident_data.iterrows():
        key = row['DR Number']
        acc_details = row[cols].to_dict()
        data[key] = acc_details

    #5. Dumping JSON file into the local system
    with open(jsonFilePath, 'w', encoding='utf-8') as jsonf:
        jsonf.write(json.dumps(data, indent=4))


# 6. Running the function to make JSON file...
data = make_json(r'../data/df_modified_ver2.csv', r'../data/traffic_data.json')

# # #7. Make request to firebase database using the request module
# with open(r'../data/traffic_data.json') as data_file:
#     data = json.load(data_file)
#     requests.put(url='https://dsci551-c6ed4-default-rtdb.firebaseio.com/traffic.json', json=data)

In [49]:
# #7. Make request to firebase database using the request module - Not working
with open(r'../data/traffic_data.json') as data_file:
    data = json.load(data_file)
    requests.put(url='https://dsci551-c6ed4-default-rtdb.firebaseio.com/traffic.json', json=data)